### ISSUE TITLE
[question] static code analysis over multiple packages

### ISSUE THREAD

**Issue by AneGris** [conan-io/conan#10708]
Hello,

tl;dr: our goal is to run static code analysis over all code used in our final application from multiple component packages, but we don't know how to get the code.

Some background: we are coming from an old monolithic build approach and are looking to migrate to a more modular approach using Conan and CMake. In the old world, we had a long list of files to feed into the compiler, a list of include directories and a list of compiler switches. From there it was rather trivial to re-use these lists as an input for the static code analysis.

We have split our monolith into several components that are each pre-built and packaged with Conan. Then we have an integration recipe for the main app, that "requires" all of the relevant components and some glue code to put it all together. So if you want to build the main app, you just need the integration recipe and the glue code. Since the components are pre-built, Conan will just download the binaries and can build the final app. This is fast and great for the purpose of the regular build. However, the exported sourcecode isn't even downloaded, thus there is no chance to access any of it for the static code analysis.

To add some complexity: our component recipes typically have options which are handed down to CMake to build variants of the components. The options can be used to determine which of the exported .c files are actually used during the build. Since this is done in CMake, Conan does not know which of the files will actually be used. The final app also has many variants that might "require" different components and set different options for these components. Thus, a specific instance of the final app will consist of a certain subset of .c code (and compiler switches). Only that subset should be considered for the static code analysis.

We have thought about just packaging the sourcecode alongside the binaries and importing that in the integration recipe. However, this does not cover the variants described in the last paragraph, unless we move/duplicate the code from the CMake file back into the recipe. Not sure how to handle compiler switches. Also, from our limited experience with Conan, it feels redundant and "incorrect" (but we could be wrong). As described in the beginning, the call of the compile toolchain in the old world was very similar to calling the static code analysis, which steered our minds towards writing our own custom generator for this, but that custom generator would need to be synched to what is used in the CMake build and overall is a daunting endeavour for Conan newbies such as ourselves.

If you've made it this far, thanks for reading! We would love to hear your thoughts!

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @AneGris 

I think there are 2 different aspects necessary to address that use case:

> We have split our monolith into several components that are each pre-built and packaged with Conan. Then we have an integration recipe for the main app, that "requires" all of the relevant components and some glue code to put it all together. So if you want to build the main app, you just need the integration recipe and the glue code. Since the components are pre-built, Conan will just download the binaries and can build the final app. This is fast and great for the purpose of the regular build. However, the exported sourcecode isn't even downloaded, thus there is no chance to access any of it for the static code analysis.

The first aspect is where is the "original" source code, how it is used by Conan, how can it be retrieved by Conan. There are different ways to achieve this, like using ``exports_sources``, using ``scm``, packaging from another repo and use the ``source()`` method, etc. What approach are you following?

> Also, from our limited experience with Conan, it feels redundant and "incorrect" (but we could be wrong). As described in the beginning, the call of the compile toolchain in the old world was very similar to calling the static code analysis, which steered our minds towards writing our own custom generator for this, but that custom generator would need to be synched to what is used in the CMake build and overall is a daunting endeavour for Conan newbies such as ourselves.

Is it not possible to do the static analysis per package? That would fit more naturally of course, and it could be driven by some settings or options, being very integrated in the flow, like ``conan create . --something=analysis`` would be creating a fake "binary" that is the static analysis one.

If the static analysis must be done necessarily always from the downstream consumer application, then it would be necessary that every package contains the information that maps at least the components to the source code. When ``build()`` happens, it could generate a "map" {component->list of source code files}, containing the specific map for the component and for this specific binary configuration. That map can be stored in a json file, and could be packaged and read by ``package_info()``, which will load it even if the package is not being built from sources and only reused from binary. Aggregating all the information for dependencies from the consumer via ``self.dependencies`` seems doable. This would still be decoupled from the actual location of sources that would be the model to obtain the list of all files involved in a given application executable.

**Comment by AneGris** [comment]
Hi @memsharded,

thanks for your response!

> The first aspect is where is the "original" source code, [...]

For the moment we have put the recipes alongside the "original" source code and are using `exports_sources`. Our source code repo structure is still not final. I'm working under the assumption that it will challenging to setup the source code repo(s) to reflect whatever is in a specific variant/version of the final software.

> Is it not possible to do the static analysis per package? [...]

While we will need to run an analysis per package, we also need to perform checks on the complete software, unfortunately.

> [...] When `build()` happens, it could generate a "map" {component->list of source code files}, containing the specific map for the component and for this specific binary configuration. That map can be stored in a json file, and could be packaged and read by `package_info()`, which will load it even if the package is not being built from sources and only reused from binary.

From what I understand, we would need to do this from within CMake, as this is currently the place where it is finally decided, which c-Files go into the final component. It appears to me that CMake is not very good at writing json, but I guess we could clobber that together using string operations. Maybe a more trivial format could be used and we can do the parsing in a python function in the recipe. We could then put that into something like `self.user_info.cFiles` in the `package_info()` method. Same for include folders and compiler switches. Or do you have a better idea? Maybe some of the files generated by CMake could be parsed; I will look into that.

Re-reading your post, I think you are suggesting to package the file itself and let the consumer deal with it? So maybe we would have something like `self.user_info.staticAnalysisJson = foo.json`. The question how to best generate this file in the first place remains.

> Aggregating all the information for dependencies from the consumer via `self.dependencies` seems doable.

This part I am struggling to understand. Are you referring to usage in a custom generator? Or is this something you could do directly in (what I called) the integration recipe?

> This would still be decoupled from the actual location of sources that would be the model to obtain the list of all files involved in a given application executable.

I don't think I follow, could you please elaborate a bit more on this point?

My fear is that if we naively create the json as described above, we would get some references pointing to the .c files in the build directory in the Conan cache. But from my understanding this directory would not exist, if a pre-built binary is downloaded. Hence the idea in the initial post of packaging the .c files themselves. We would then of course need to "remap" the map file, which should be feasible.

Am I on the right track or am I misunderstanding the ideas you presented?

**Comment by memsharded** [comment]
Sorry this was not responded or followed up back then.

Marking it as staled, to be closed soon, but please let us know if there are still questions or need any help.

**Comment by memsharded** [comment]
Closing as staled, please comment and we will re-open or create new tickets as necessary, thanks for the feedback!
